public/app/features/explore/state/query.ts

```diff
@@ -607,51 +615,115 @@ export const runQueries = (
             datasourceInstance,
-            type,
-            {
-              ...transaction.request,
-              requestId: `${transaction.request.requestId}_${snakeCase(type)}`,
-            },
-            newQuerySource
-          );
-
-          if (dataProvider) {
-            dispatch(
-              storeSupplementaryQueryDataProviderAction({
-                exploreId,
-                type,
-                dataProvider,
-              })
-            );
-
-            if (!canReuseSupplementaryQueryData(supplementaryQueries[type].data, queries, absoluteRange)) {
-              dispatch(cleanSupplementaryQueryAction({ exploreId, type }));
-              if (supplementaryQueries[type].enabled) {
-                dispatch(loadSupplementaryQueryData(exploreId, type));
-              }
-            }
-          } else {
-            // If data source instance doesn't support this supplementary query, we clean the data provider
-            dispatch(
-              cleanSupplementaryQueryDataProviderAction({
-                exploreId,
-                type,
-              })
-            );
-          }
-        }
+            transaction,
+            newQuerySource,
+            supplementaryQueries,
+            queries,
+            absoluteRange,
+          })
+        );
       }
     }
 
     dispatch(queryStoreSubscriptionAction({ exploreId, querySubscription: newQuerySubscription }));
   };
 };
 
+type HandleSupplementaryQueriesOptions = {
+  exploreId: ExploreId;
+  transaction: QueryTransaction;
+  datasourceInstance: DataSourceApi;
+  newQuerySource: Observable<ExplorePanelData>;
+  supplementaryQueries: SupplementaryQueries;
+  queries: DataQuery[];
+  absoluteRange: AbsoluteTimeRange;
+};
+
+const groupDataQueries = async (queries: DataQuery[], scopedVars: ScopedVars) => {
+  queries = queries.filter((t) => {
+    return t.datasource?.uid !== MIXED_DATASOURCE_NAME;
+  });
+  // Build groups of queries to run in parallel
+  const sets: { [key: string]: DataQuery[] } = groupBy(queries, 'datasource.uid');
+  const mixed: Array<{ datasource: DataSourceApi; targets: DataQuery[] }> = [];
+
+  for (const key in sets) {
+    const targets = sets[key];
+    mixed.push({
+      datasource: await getDataSourceSrv().get(targets[0].datasource, scopedVars),
+      targets,
+    });
+  }
+
+  return mixed;
+};
+
+const handleSupplementaryQueries = createAsyncThunk(
+  'explore/handleSupplementaryQueries',
+  async (
+    {
+      datasourceInstance,
+      exploreId,
+      transaction,
+      newQuerySource,
+      supplementaryQueries,
+      queries,
+      absoluteRange,
+    }: HandleSupplementaryQueriesOptions,
+    { dispatch }
+  ) => {
+    let datasources;
+    if (datasourceInstance.meta.mixed) {
+      datasources = await groupDataQueries(transaction.request.targets, transaction.request.scopedVars);
+    } else {
+      datasources = [{ datasource: datasourceInstance, targets: transaction.request.targets }];
+    }
```

Comment:
can we use a more accurate name for the `datasources` variable?
something like `queriesByDatasource` or `groupedQueries` maybe?

Reply:
Updated in 5bd2458a1f921a13e432599af119655747cd6120.